### PR TITLE
Initial documentation for learn, gensfen, convert, and binpack.

### PR DESCRIPTION
**Pull request by Sopel97**
This is the first step to having well documented learn, gensfen, convert commands. Currently I documented their general purpose, invocation syntax, and briefly described parameters. Further work could focus on documenting recommended settings.

I allowed myself to skip writing documentation for shuffle and old convert stuff. Partially because I don't understand it well and partially because I consider it on its way out. What to do with this can be tackled in further PRs.

I also took a comment from one of old issues briefly describing the binpack format in put it in `docs/binpack.md` so that it's not completely forgotten.

Some things that I would like others to check and improve upon before merging this:
- `bat` and `minibatch` options in `learn` seems to be about the same thing. Also I don't quite understand the difference between this batch size and `nn_batch_size` so the docs around that are lacking.
- I have no idea what `nn_options` could be used for.
- `winning_probability_coefficient` is a mystery.
- I marked some [old] parameter name aliases as deprecated, for some I wasn't sure which one should stay though.

This may also be a good time to rethink some defaults...

### CODE REVIEW

File: src/docs/convert.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Convert
+
+`convert` allows conversion of training between any of `.plain`, `.bin`, and `.binpack`.
```
Comment (vondele) [inline src/docs/convert.md:3]: training data

File: src/docs/convert.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+The syntax of this command is as follows:
+```
+convert from_path to_path [append]
+```
+
+`from_path` is the path to the file to convert from. The type of the data is deduced based on its extension (one of `.plain`, `.bin`, `.binpack`).
+`to_path` is the path to an output file. The type of the data is deduced from its extension. If the file does not exist it is created.
+
+Last argument is optional. If not specified then they output file will be truncated prior to any writes. If the last argument is `append` then the converted training data will be appended to the end of the output file.
```
Comment (vondele) [inline src/docs/convert.md:15]: The last argument ...
the they output -> then the output

File: src/docs/gensfen.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+`depth2` - maximum depth of evaluation of each position. If not specified then the same as `depth`.
+
+`nodes` - the number of nodes to use for evaluation of each position. This number is multiplied by the number of PVs of the current search. This does NOT override the `depth` and `depth2` options. If specified then whichever of depth or nodes limit is reached first applies.
+
+`loop` - the number of training data entries to generate. 1 entry == 1 position.
+
+`output_file_name` - the name of the file to output to. If the extension is not present or doesn't match the selected training data format the right extension will be appened.
+
+`eval_limit` - evaluations with higher absolute value than this will not be written and will terminate a self-play game. Should not exceed 10000 which VALUE_KNOWN_WIN, but is only hardcapped at mate in 2 (\~30000).
```
Comment (vondele) [inline src/docs/gensfen.md:21]: which is VALUE_KNOWN_WIN

File: src/docs/gensfen.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+`write_out_draw_game_in_training_data_generation` - deprecated, alias for `use_draw_in_training_data_generation`
+
+`detect_draw_by_consecutive_low_score` - either 0 or 1. If 1 then drawn games will be adjudicated when the score remains 0 for at least 8 plies after ply 80. Default: 0.
+
+`use_game_draw_adjudication` - deprecated, alias for `detect_draw_by_consecutive_low_score`
+
+`detect_draw_by_insufficient_mating_material` - either 0 or 1. If 1 then position with insufficient material will be adjudicated as draws. Default: 0.
+
+`sfen_format` - format of the training data to use. Either `bin` or `binpack`. Default: `bin`.
```
Comment (vondele) [inline src/docs/gensfen.md:55]: should probably default to binpack..

File: src/docs/learn.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
+# Learn
+
+`learn` command allows allows training a network from training data.
```
Comment (vondele) [inline src/docs/learn.md:3]: allows allows -> allows

File: src/docs/learn.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+As all commands in stockfish `learn` can be invoked either from command line (as `stockfish.exe learn ...`, but this is not recommended because it's not possible to specify UCI options before `learn` executes) or in the interactive prompt.
+
+`learn` takes named parameters in form `learn param_1_name param_1_value param_2_name param_2_value ...`. Unrecognized parameters form a list of paths to training data files.
```
Comment (vondele) [inline src/docs/learn.md:7]: in the form

File: src/docs/learn.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+`newbob_decay` - learning rate will be multiplied by this factor every time a net is rejected (so in other words it controls LR drops). Default: 1.0 (no LR drops)
+
+`newbob_num_trials` - determines after how many subsequent rejected nets the training process will be terminated. Default: 2.
+
+`nn_options` - if you're reading this you don't use it. It passes messages directly to the network evaluation. I don't know what it can do either.
+
+`eval_save_interval` - every `eval_save_interval` positions the network will be saved and either accepted or rejected (in which case an LR drop follows). Default: 1000000000 (1B). (generally people use values in 10M-100M range)
+
+`loss_output_interval` - every `loss_output_interval` fittness statistics are displayed. Default: `batchsize`
```
Comment (vondele) [inline src/docs/learn.md:63]: fitness

**Comment by vondele** [comment]
I agree that having a reasonable example to train a net would be valuable (I still don't know how to do it ;-)), and from a user point of view to have the defaults such that this reasonable example is as short as possible.

File: src/docs/learn.md (hunk 6 above)
Comment (Sopel97) [reply src/docs/learn.md:7]: I think in this case "form" is correct because it meant as in "Unrecognized parameters are parsed as paths and are added to the list of training data files". Though maybe I should use that phrasing?

Comment (vondele) [reply src/docs/learn.md:7]: I was talking about the first 'form' i.e. `takes named parameters in the form`  (but I'm no native speaker). The second form is correct.

File: src/docs/gensfen.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
+# Gensfen
+
+`gensfen` command allows generation of training data from self-play in a manner that suits training better than traditional games. It introduces random moves to diversify openings, allows reduced pruning, disabling of TT for less interference between searches, and fixed depth evaluation.
```
Comment (nodchip) [inline src/docs/gensfen.md:3]: I recommend to **enable** TT for gensfen.  Because disabling TT in gensfen slows down the speed.  I recommend to **disable** TT for learn for less interference between searches.

File: src/docs/gensfen.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+`write_minply` - minimum ply for which the training data entry will be emitted.
+
+`write_maxply` - maximum ply for which the training data entry will be emitted.
+
+`save_every` - the number of training data entries per file. If not specified then there will be always one file. If specified there may be more than one file generated (each having at most `save_every` training data entries) and each file will have a unique number attached.
+
+`random_file_name` - if specified then the output filename will be chosen randomly. Overrides `output_file_name`.
+
+`use_draw_in_training_data_generation` - either 0 or 1. If 1 then training data from drawn games will be emitted too. Default: 0.
```
Comment (nodchip) [inline src/docs/gensfen.md:45]: `use_draw_in_training_data_generation` is a deprecated name, and is the alias for `use_draw_in_training_data_generation`.

File: src/docs/gensfen.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
+# Gensfen
+
+`gensfen` command allows generation of training data from self-play in a manner that suits training better than traditional games. It introduces random moves to diversify openings, allows reduced pruning, disabling of TT for less interference between searches, and fixed depth evaluation.
+
```
Comment (nodchip) [inline src/docs/gensfen.md:4]: Could you please describe that it is recommended to set `PruneAtShallowDepth` to false.

File: src/docs/learn.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
+# Learn
+
+`learn` command allows allows training a network from training data.
+
+As all commands in stockfish `learn` can be invoked either from command line (as `stockfish.exe learn ...`, but this is not recommended because it's not possible to specify UCI options before `learn` executes) or in the interactive prompt.
+
```
Comment (nodchip) [inline src/docs/learn.md:6]: Could we describe below?
- It is recommended to set `EnableTranspositionTable` to `false`.
- It is recommended to set `PruneAtShallowDepth` to `false`.
- It is needed to set `Use NNUE` to `pure`.

File: src/docs/learn.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+`bat` - the size of a minibatch in multiples of 10000. The number of positions inbetween weights updates. Default: 1000 (meaning mini batch size of 1000000).
+
+`targetdir` - path to the direction from which training data will be read. All files in this directory are read sequentially. If not specified then only the list of files from positional arguments will be used. If specified then files from the given directory will be used after the explicitly specified files.
+
+`loop` - the number of times to loop over all training data.
+
+`basedir` - the base directory for the paths. Default: "" (current directory)
+
+`batchsize` - same as `bat` but doesn't scale by 10000
```
Comment (nodchip) [inline src/docs/learn.md:19]: The trainer reads `batchsize` entries at a time from training data, shuffles them, and uses each `nn_batch_size` entries as minibatch.

**Review by nodchip — CHANGES_REQUESTED** [review]
Thank you for writing documents.  I added some comments.  Could you please take a look at them?

**Comment by nodchip** [comment]
The changes looks good to me.  I will merge this pull request after @vondele 's review is finished.

**Comment by Sopel97** [comment]
So you mean we adjust the defaults in this PR?

**Comment by nodchip** [comment]
I don't think that we should adjust the defaults in this PR.  It also looks that all the points other than default values pointed out by @vondele are finished.  I will merge this PR.
